File: lib/assets/javascripts/farm.js
// Constructor
function Farm(name, location, boundaries) {
  this.init(name, location, boundaries)
}
  
Farm.prototype.init = function (name, location, boundaries) {
  var myself = this;

  this.name = name;
  this.location = { 
    latitude: location.latitude,
    longitude: location.longitude
  };

  this.bounds = {
    north: location.latitude,
    south: location.latitude,
    east:  location.longitude,
    west:  location.longitude
  };

  this.fields = [];

  // Find the bounds
  this.boundary = [];
  boundaries.forEach( function(boundary) {
    boundary.forEach( function(coord) {

      // determine farm's bounding box
      if(coord.latitude > myself.bounds.north)
        myself.bounds.north = coord.latitude;
      if(coord.latitude < myself.bounds.south)
        myself.bounds.south = coord.latitude;
      if(coord.longitude > myself.bounds.east)
        myself.bounds.east = coord.longitude;
      if(coord.longitude < myself.bounds.west)
        myself.bounds.west = coord.longitude;

    });
    myself.fields.push(new Field(boundary));
  }); 

  // 111,111 meters = 1 degree latitude
  this.height = Math.ceil(111111 * (this.bounds.north - this.bounds.south));
  degreesPerMeterLat = 1/111111;

  // 111,111 * cos(latitude) = 1 degree longitude
  this.width = Math.ceil(111111 * Math.cos(this.location.latitude * (Math.PI/180)) * (this.bounds.east - this.bounds.west));
  degreesPerMeterLng = 1/(111111 * Math.cos(this.location.latitude * (Math.PI/180)));

  console.log("width x height", this.width, this.height);
  // Create the patches
  this.patches = new Array(this.width * this.height);
//  console.log("HERE");
  for(i = 0; i < this.width; i++){
    for(j = 0; j < this.height; j++){
      this.patches[i + j] = new Patch(
        this.bounds.west + (i * degreesPerMeterLat),
        this.bounds.north + (j * degreesPerMeterLng)
      );
//      console.log(this.bounds.west + i * degreesPerMeterLat, this.bounds.north + j * degreesPerMeterLng);
//      this.patches[j * this.width + i] = new Patch({
//        latitude: this.bounds.west + i * degreesPerMeterLat,
//        longitude: this.bounds.north + j * degreesPerMeterLng
//      });
    }
  }
//  console.log(this.patches[40])
  
};

File: app/models/farm.rb
class Farm < ActiveRecord::Base
  has_paper_trail

  belongs_to :user

  validates_presence_of :user, :name, :latitude, :longitude, :data

  def to_svg
    @svg ||= generate_svg
  end

  def static_map_url
    bounds = JSON.parse(data)["field_bounds"] || []
    bounds.reject!(&:empty?)
    paths = bounds.collect do |field| 
      "path=color:0x000000ff|#{(field.collect{ |corner| "#{ corner["latitude"] },#{ corner["longitude"] }"} + ["#{field[0]["latitude"]},#{field[0]["longitude"]}"] ).join("|") }".html_safe
    end
    args = ["http://maps.googleapis.com/maps/api/staticmap?", #key=AIzaSyAWk3UQFfLP1-q-VrLd46sKcZ-RyYMDfTM",
       "markers=color:red|#{latitude},#{longitude}",
       "zoom=13",
       "maptype=terrain",
       "size=600x300"
      ]
     (args + paths).join("&")
  end



private

  def generate_svg
    width = 300
    height = 200
    svg = "<svg viewbox='0 0 #{width} #{height}'>"

    farm_data = JSON.parse data
    north = farm_data["location"]["latitude"]
    south = farm_data["location"]["latitude"]
    east = farm_data["location"]["longitude"]
    west = farm_data["location"]["longitude"]

    if farm_data["field_bounds"]

      # Find the outer boundaries of the simulation
      for bound in farm_data["field_bounds"]
        for corner in bound
          north = corner["latitude"] > north ? corner["latitude"] : north;
          south = corner["latitude"] < south ? corner["latitude"] : south;
          west = corner["longitude"] > west ? corner["longitude"] : west;
          east = corner["longitude"] < east ? corner["longitude"] : east;
        end
      end

      # Convert field bounds into pixel coordinates
      farm_data["field_bounds"].each_with_index do |bound, index|
        points = []
        left = width
        right = 0
        top = 0
        bottom = height
        for corner in bound
          x = (corner["longitude"] - east) * (width / (west - east))
          y = height - (corner["latitude"] - south) * (height / (north - south))
          left = x < left ? x : left
          right = x > right ? x : right
          top = y > top ? y : top
          bottom = y < bottom ? y : bottom
          points << "#{x},#{y}"
        end
        svg << "<polygon data-field-index='#{ index }' points='#{ points.join(" ") }' style='stroke:black;stroke-width:5' />"
        svg << "<text x='#{left + ((right - left) / 2)}' y='#{bottom + ((top - bottom) / 2)}' font-family='Veranda' font-size='55'>#{ index }</text>"
      end
    end
    svg << "</svg>"
  end

end
